[Constraint solver] Do argument label matching during apply simplification

DougGregor opened the pull request:
When simplifying a function application constraint, check the argument
labels for that application against the disjunction containing the overload
set, disabling any overloads with mis-matching labels. This is staging for
several different directions:

* Eliminating the argument label matching from performMemberLookup, where it
does not belong
* More aggressively filtering the overload set when we have some concrete
information about argument types
* Identifying favored constraints when we have some concrete information
about argument types

At present, the only easily-visible effect of this change is that
we now properly handle argument label matching for non-member functions,
subscripts, and unresolved member references (e.g., `.foo(bar: 0)`). It also
improves the common-result-type computation a bit, although the effect of
that is harder to see.

DougGregor committed:
[Constraint solver] Match call arguments for subscripts.

Record the argument labels provided to a subscript in the solver, and
use that to filter out subscript declarations with non-matching
argument labels during function application. This reduces the number of
overloaded subscript declarations that will be considered in later
steps in the solver, reducing the solution space.

*Disable* this optimization in the normal member-lookup path, which is
intended to go away in the near future. This limits the scope of the
change somewhat, so we can separately tackle the diagnostics issue.
The one diagnostics change here is probably an improvement, because
the user explicitly stated the argument labels, and is more likely
missing a conversion on the argument than having typed the wrong
label.

DougGregor committed:
[Constraint solver] Do argument label matching during apply simplification.

When simplifying a function application constraint, check the argument
labels for that application against the disjunction containing the overload
set, disabling any overloads with mis-matching labels. This is staging for
several different directions:

* Eliminating the argument label matching from performMemberLookup, where it
does not belong
* More aggressively filtering the overload set when we have some concrete
information about argument types
* Identifying favored constraints when we have some concrete information
about argument types

At present, the only easily-visible effect of this change is that
we now properly handle argument label matching for non-member functions.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test compiler performance

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Constraint solver] Restrict key path application to subscript[keyPath:].

Extend argument label matching to handle key path application choices,
which are accessed via subscript syntax with the argument labels “keyPath:”.
This gets us to the point where it’s possible to filter down to a single
subscript if there is no type-based overloading going on.

DougGregor committed:
[Constraint solver] Declarations with IUOs don’t have effective overload types

A declaration with an implicitly-unwrapped optional essentially has two
effective overload types, because the result might be optional or it might
have been forced. Disable computation of the effective overload type in this
case.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
Update test case for subscript argument labels

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+        // We must be binding the type variable (or a type variable
+        // equivalent to it).
+        auto boundTypeVar = constraint->getFirstType()->getAs<TypeVariableType>();
+        if (!boundTypeVar || getRepresentative(boundTypeVar) != fnTypeVar) {
```
Should both of these checks be assertions since `getUnboundBindOverloadDisjunction` is supposed to return only viable `bind` disjunctions associated with current type variable?

DougGregor replied:
`getUnboundBindOverloadDisjunction()` isn't quite doing all of that checking, but it looks like other clients expect this, too. I'll update it accordingly and simplify `filterDisjunction()`.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+
+  // Consider each of the constraints in the disjunction.
+retry_after_fail:
+  bool skippedAnyConstraints = false;
```
Maybe `hasAmbiguousChoices` or something like that?

DougGregor replied:
`hasUnhandledConstraints` is maybe better. It's less about "ambiguous" and more about "we couldn't handle a constraint so we have to be conservative"

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    return fnType;
+
+  // If we have a common result type, bind the expected result type to it.
+  if (commonResultType && *commonResultType) {
```
This seems a bit weird to me - should setting an optional with empty/invalid type be avoided?

DougGregor replied:
I need three states: not computed, computed, and conflicting. I could use a `Type` with a different sentinel (e.g., `ErrorType`), I guess.

xedin replied:
I think it might make sense to use Type here for this, since this check means - computed but empty/invalid)?

DougGregor replied:
Without the optional, it'll mean doing a `commonResultType && !commonResultType->is<ErrorType>()` here, which is only slightly better.

xedin replied:
I don’t have much preference here, but maybe UresolvedType would be more appropriate to indicate conflict. You can leave it as optional here, it just seems like a quirk of C++ where you can have optional of null pointer...

DougGregor replied:
`UnresolvedType` is something I'd like us to remove, so I'd rather not create a dependency on it. I'll go with `ErrorType`.

xedin replied:
Sure!

xedin reviewed (commented):
Overall LGTM! I've left some comments inline.

DougGregor commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3557,7 +3583,13 @@ performMemberLookup(ConstraintKind constraintKind, DeclName memberName,
+        (!candidate.isDecl() || !isa<SubscriptDecl>(candidate.getDecl())) &&
+        !areConservativelyCompatibleArgumentLabels(
+            candidate, argumentLabels->Labels,
+            argumentLabels->HasTrailingClosure)) {
```
Getting there! I want to make sure it's working, first.

xedin replied:
Sounds good!

DougGregor commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
 }
 
+Type ConstraintSystem::simplifyAppliedOverloads(
+                                Type fnType,
```
Common type computation will eventually put more structure here (e.g., `($T0, $T0) -> $T0`)), which is why I left it as a type (since this will be the signature we expect). I could make it `TypeVariableType` now and generalize later, sure.

DougGregor replied:
Yeah, I'll make it `TypeVariableType` for now to be more clear.

DougGregor commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -3064,6 +3109,30 @@ class ConstraintSystem {
+  /// \returns One of \c Solved (only a single term remained),
+  /// \c Unsolved (more than one disjunction terms remain), or
+  /// \c Error (all terms were filtered out).
+  SolutionKind filterDisjunctions(Constraint *disjunction,
```
Will do!

DougGregor committed:
[Constraint solver] Address feedback from Pavel and Slava.

DougGregor committed:
[Constraint solver] Synchronize argument label setting/retrievable.

The walker that was setting argument labels was looking through ! and ?
postfix expressions, but the lookup code itself was not, leading to missed
opportunities for filtering based on argument labels. Generalize the
transformation that maps from the function expression down to the locator
for which we will retrieve argument labels.

DougGregor committed:
[Constraint solver] Fix effective-overload-type for constructors, dynamic Self

Implement support for querying the effective overload type for constructors
and fix a semi-related bug for methods returning dynamic Self, which I
had not accounted for.

DougGregor committed:
[Constraint solver] Minor cleanup for @optional declarations.

DougGregor committed:
[Constraint solver] Match argument labels for unresolved member expressions.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please test compiler performance

DougGregor committed:
The test case from rdar://problem/20818064 is now scaling well.

DougGregor commented:
The macOS failure is a slow test case (from rdar://problem/20818064) that is no longer taking exponential time. My upcoming commit will move the test case from "slow" to "fast" and bump up the scaling factor to be sure we've fixed it.

DougGregor committed:
Add a "slow" type checker performance test for rdar://problem/46713933

DougGregor committed:
Use ErrorType as a sentinel type rather than Optional<Type>

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test
